pipeline/src/pipeline.rs

```diff
@@ -874,16 +864,28 @@ impl<T: FieldElement> Pipeline<T> {
                     .backend
                     .expect("backend must be set before generating verification key!");
                 let factory = backend.factory::<T>();
-                let backend = if let Some(path) = self.arguments.setup_file.as_ref() {
-                    let mut file = fs::File::open(path).unwrap();
-                    factory.create_from_setup(&mut file).unwrap()
-                } else {
-                    factory.create(pil.degree())
-                };
 
-                match backend.verification_key(pil, fixed_cols) {
-                    Ok(vkey) => Ok(vkey),
-                    Err(powdr_backend::Error::VerificationKeyFailed(e)) => Err(vec![e]),
+                let mut setup_file = self
+                    .arguments
+                    .setup_file
+                    .as_ref()
+                    .map(|path| BufReader::new(fs::File::open(path).unwrap()));
+
+                let backend = factory
+                    .create(
+                        pil.borrow(),
+                        &fixed_cols[..],
```

Comment:
In general I find it annoying to pass the fixed_cols around like this (which is not this PR's fault). We should have some type "low level program" that has the PIL constraints AND the fixed columns.